### PR TITLE
new: Add in ability to create/update/delete MISP Event Reports

### PR DESCRIPTION
**Pull request by tomking2**
Now we have Event Reports in MISP, thought it would be a good thing to be able to manage this through PyMISP.

Unfortunately with the current MISP codebase, there are a few limitations:
- The 'hard' delete flag must be passed in the url as `/1` and not as a POSTed JSON.
- This must be managed through the PyMISP func, e.g. `pm.add_event_report` and not through the `MISPEvent` class. MISP doesn't seem to recognise changes to `EventReports` and update accordingly when POSTed in the Event JSON. As such, I have not included a setter prop for the `MISPEvent.event_reports property.`

Does what we're looking for though :)

### CODE REVIEW

**Comment by Rafiot** [comment]
Looks good, let me know when I should merge it.

Also, would you mind adding a test in the test suite (https://github.com/MISP/PyMISP/blob/main/tests/testlive_comprehensive.py)? Otherwise, I'll do it

**Comment by tomking2** [comment]
@Rafiot Great :). Yeah I'll add one in shortly. Once updated, should be good for you to merge.

**Comment by tomking2** [comment]
@Rafiot, nosetest in place and confirmed pass on local instance, should be ready for your final lookover.

File: pymisp/api.py
Comment (Rafiot) [inline pymisp/api.py:474]: I haven't tested yet, but does MISP returns a MISPEventReport on delete?

File: tests/testlive_comprehensive.py
Comment (Rafiot) [inline tests/testlive_comprehensive.py:2608]: It doesn't seems to return an EventReport there.

**Review by Rafiot — CHANGES_REQUESTED** [review]
Looks good to me. We may want to add a few more helpers to assist the creation of reports later, but it looks pretty good. 
I think the only change is the return of `delete_event_report`.

**Comment by Rafiot** [comment]
From PyMISP perspective, it looks good (after changing the delete method). @mokaddem if you have any further comments, go for it.

File: pymisp/api.py
Comment (tomking2) [reply pymisp/api.py:474]: MISP does return a MISPEventReport on soft delete. For a hard delete it does not (see line 475 for the addition of the `hard` param check

File: tests/testlive_comprehensive.py
Comment (tomking2) [reply tests/testlive_comprehensive.py:2608]: That'll be the `hard` delete for line 2608-2609. Line 2604-2606 should return an EventReport with the updated `deleted` flag

File: pymisp/api.py
Comment (Rafiot) [reply pymisp/api.py:474]: alright, that's weird, and inconsistent with the other delete calls. @mokaddem is it expected?

Comment (tomking2) [reply pymisp/api.py:474]: Any update on this one @mokaddem  CC @Rafiot

Comment (mokaddem) [reply pymisp/api.py:474]: @Rafiot I'm confused about the inconsistency you are mentioning.

Currently, **soft** deleting an Event report returns the Event report blob:
```json
{
    "EventReport": {
        "id": "x",
        "deleted": true
        ...
    }
}
```

**Hard** deleting a report returns the generic REST response:
```json
{
    "saved": true,
    "success": true,
    "name": "Event Report X hard deleted",
    "message": "Event Report X hard deleted",
    "url": "\/event_reports\/delete\/X"
}
```
What else would you expect? It feels weird to return the event report data as it has been deleted on the server.
For example, **hard** deleting an attribute returns the following:
```json
{
    "message": "Attribute deleted."
}
```

**Comment by mokaddem** [comment]
> The 'hard' delete flag must be passed in the url as /1 and not as a POSTed JSON.

This is fixed in b36f685d88b8c91e6b5879445cb29363b4231157
I'm gonna have a look at the second point you mentioned

File: pymisp/api.py
Comment (Rafiot) [reply pymisp/api.py:474]: Soft deleting something on MISP doesn't return the thing (at least it doesn't for attributes and events). It's not really a problem if deleting an event report does so, it's just different from the other calls.

**Comment by mokaddem** [comment]
@tomking2 Your second point was a nice catch:
> [...] MISP doesn't seem to recognise changes to EventReports and update accordingly when POSTed in the Event JSON [...]

And it's fixed in https://github.com/MISP/MISP/commit/b86c58c8ae24747da06878de178de1cb067a53e0

File: pymisp/api.py
Comment (mokaddem) [reply pymisp/api.py:474]: Ah the inconsistency was the other way around. Didn't get that.
It's changed https://github.com/MISP/MISP/commit/5e6444a663085c2f45e33fed2240f5e17e484b36

Comment (Rafiot) [reply pymisp/api.py:474]: Thank you!

**Comment by tomking2** [comment]
> @tomking2 Your second point was a nice catch:
> 
> > [...] MISP doesn't seem to recognise changes to EventReports and update accordingly when POSTed in the Event JSON [...]
> 
> And it's fixed in [MISP/MISP@b86c58c](https://github.com/MISP/MISP/commit/b86c58c8ae24747da06878de178de1cb067a53e0)

Would you look at that :), cheers @mokaddem!

Should be OK to merge in if I'm not mistaken @Rafiot

**Comment by Rafiot** [comment]
I'll have a look tomorrow, and figure out the failures on github actions.

**Review by Rafiot — APPROVED** [review]

**Pull request merged**
